core/src/main/java/com/linecorp/armeria/server/cors/AbstractCorsPolicyBuilder.java

```diff
@@ -0,0 +1,383 @@
+     * @param values the values for the HTTP header.
+     * @return {@link B} to support method chaining.
+     */
+    public B preflightResponseHeader(final AsciiString name, final Object... values) {
+        requireNonNull(name, "name");
+        requireNonNull(values, "values");
+        for (int i = 0; i < values.length; i++) {
+            if (values[i] == null) {
+                throw new NullPointerException("values[" + i + ']');
+            }
+        }
+        preflightResponseHeaders.put(name, new ConstantValueSupplier(Arrays.asList(values)));
+        return self();
+    }
+
+    /**
+     * Returns HTTP response headers that should be added to a CORS preflight response.
+     *
+     * <p>An intermediary like a load balancer might require that a CORS preflight request
+     * have certain headers set. This enables such headers to be added.
+     *
+     * @param name the name of the HTTP header.
+     * @param values the values for the HTTP header.
+     * @param <T> the type of values that the Iterable contains.
+     * @return {@link B} to support method chaining.
+     */
+    public <T> B preflightResponseHeader(final AsciiString name, final Iterable<T> values) {
+        requireNonNull(name, "name");
+        requireNonNull(values, "values");
+        preflightResponseHeaders.put(name, new ConstantValueSupplier(values));
+        return self();
+    }
+
+    /**
+     * Returns HTTP response headers that should be added to a CORS preflight response.
+     *
+     * <p>An intermediary like a load balancer might require that a CORS preflight request
+     * have certain headers set. This enables such headers to be added.
+     *
+     * <p>Some values must be dynamically created when the HTTP response is created, for
+     * example the 'Date' response header. This can be accomplished by using a {@link Supplier}
+     * which will have its 'call' method invoked when the HTTP response is created.
+     *
+     * @param name the name of the HTTP header.
+     * @param valueSupplier a {@link Supplier} which will be invoked at HTTP response creation.
+     * @param <T> the type of the value that the {@link Supplier} can return.
+     * @return {@link B} to support method chaining.
+     */
+    public <T> B preflightResponseHeader(AsciiString name, Supplier<T> valueSupplier) {
+        requireNonNull(name, "name");
+        requireNonNull(valueSupplier, "valueSupplier");
+        preflightResponseHeaders.put(name, valueSupplier);
+        return self();
+    }
+
+    /**
+     * Specifies that no preflight response headers should be added to a preflight response.
+     *
+     * @return {@link B} to support method chaining.
+     */
+    public B disablePreflightResponseHeaders() {
+        preflightResponseHeadersDisabled = true;
+        return self();
+    }
+
+    /**
+     * Returns a newly-created {@link CorsPolicy} based on the properties of this builder.
+     */
+    public CorsPolicy build() {
```

Comment:
This should not be visible when a user uses `ChainedCorsPolicyBuilder`. We could make it package-local and increase the visibility at `CorsPolicyBuilder`?